src/main/shadowedFileDetector.ts

```diff
@@ -0,0 +1,101 @@
+import path from 'path';
+import fs from 'fs';
+import os from 'os';
+import timers from 'timers';
+
+import Electron from 'electron';
+import * as window from '@/window';
+import shadowInfo from '@/utils/pathCheck';
+import resources from '@/resources';
+
+export class PathConflictManager {
+  protected pathConflicts: Record<string, Array<string>> = {};
+
+  constructor() {
+    this.pathConflicts = {};
+  }
+
+  /**
+   * Can be called either from the UI via an IPCRenderer event, or from the file-system watcher
+   * When called from the file-system watcher `event` will be undefined.
```

Comment:
It's done this way because I'm calling this function from two different contexts:
1. When the user activates a symlink button 
2. When the contents of one of the directories in the path changes

This is intended to be a lazy operation, but if a checkbox has never changed but a file of interest changes in a path directory we calculate the conflicts for that file, and they get stored if they're needed later.

Reply:
Right, the fact that it's being called from two contexts (with very distinct behaviour) was the confusing part.  I believe things would be more understandable if we just had this particular bit of code twice. (And I usually err on the side of deduplicating code…)